Implement FT.INFO command

jruaux opened the pull request:
This PR implements the FT.INFO command to retrieve index information from RediSearch.

The implementation returns a IndexInfo object.

**Changes:**
- Add IndexInfo class with index information including:
  - Basic index information (name, document counts, terms, records)
  - Index options and definition
  - Schema attributes
  - Size statistics (memory usage)
  - Indexing statistics (progress, failures, timing)
  - Garbage collection statistics
  - Cursor statistics
  - Dialect usage statistics
  - Error statistics
- Add IndexInfoParser supporting both RESP2 and RESP3 protocols
- Implement ftInfo across all API layers (sync, async, reactive, coroutines)
- Add unit tests for IndexInfo class (11 tests, all passing)
- Update integration tests

**Testing:**
All unit tests pass. Integration tests require a Redis instance with RediSearch module.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new RediSearch command and a fairly complex RESP2/RESP3 response parser with new public model types, which could affect client compatibility and parsing correctness across Redis versions.
> 
> **Overview**
> Adds support for the RediSearch `FT.INFO` command across sync/async/reactive/coroutines APIs, wiring it through `RediSearchCommandBuilder` and introducing the `FT_INFO` `CommandType`.
> 
> Introduces a new `IndexInfo` model (with nested typed stats/field definitions and a forward-compatible `additionalFields` map) plus an `IndexInfoParser` that parses both RESP2 and RESP3 responses. Adds unit tests covering option/definition/field/stat parsing scenarios and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4e904ef93cead76693365cde69b5bc6adc01ba65. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jruaux committed:
Add FT.INFO command with strongly-typed IndexInfo response

Implements FT.INFO command to retrieve index information from RediSearch.
Returns a strongly-typed IndexInfo object instead of Map<String, Object>
for better type safety and developer experience.

Changes:
- Add IndexInfo class with comprehensive index metadata
- Add IndexInfoParser supporting RESP2 and RESP3 protocols
- Implement ftInfo across all API layers (sync, async, reactive, coroutines)
- Add unit tests for IndexInfo class
- Update integration tests

jruaux committed:
Add additionalFields map to IndexInfo for forward compatibility

jruaux committed:
Fix IndexInfo type parameters: change from <K,V> to <V>

- Changed IndexInfo from IndexInfo<K, V> to IndexInfo<V>
- Index name is now always String type (decoded with StringCodec)
- Updated all ftInfo method signatures across all command interfaces
- Updated IndexInfoParser to decode index name as String
- Updated all tests to use new signature
- Removed IndexInfoParserUnitTests (integration tests provide better coverage)

All integration tests passing (testFtInfoCommand*)

jruaux committed:
Fix IndexInfoParser to handle parsed List values and field flags

- Handle case where parseListValue already parsed ComplexData into List
- Add support for String values in parseLong, parseDouble, and parseBoolean methods
- Fix parsing of field flags when SORTABLE has a flag as its value (e.g., SORTABLE NOSTEM)
- Update getBoolean to decode non-String flag values using decodeStringAsString
- Add isFieldFlag helper method to identify known field flags

tishun committed:
Polishing

jruaux commented:
@tishun Can we include this in the next release?

a-TODO-rov commented:
Hey, @jruaux 
Me and @tishun have spent some time on this PR on a multiple occasions and i personally don't think its ready to be merged.  Here are some of the problems i noticed so far:
1. This command has very complex response and its very different between RESP2/RESP3. I am concerned that this highly detailed parsing you are proposing is a huge overhead and might have some hidden bugs in it.
Good example i think are the FIELDS - especially how the flags are being parsed - [codesnippet](https://github.com/redis/lettuce/pull/3599/changes#diff-ea5b21dcf69be0d730dc392b5cb7205928e483d274f95e13a4a65b348162bf83R499-R565):
``` java
for (int i = 0; i < list.size(); i += 2) {
    if (i + 1 < list.size()) {
        String key = decodeStringAsString(list.get(i));
        Object val = list.get(i + 1);
        // Handle special case where SORTABLE has a flag as its value (e.g., SORTABLE NOSTEM)
        // In this case, we need to add both SORTABLE and the flag (NOSTEM) as separate keys
        if (FLAG_SORTABLE.equals(key)) {
            attributeMap.put(key, val);
            String valStr = decodeStringAsString(val);
            // Check if the value is a known flag
            if (isFieldFlag(valStr)) {
                attributeMap.put(valStr, valStr);
            }
        } else {
            attributeMap.put(key, val);
        }
    }
}
```
How will this function behave for this response:
``` java
 7) attributes
 8) 1)  1) identifier
        2) title
        3) attribute
        4) product_title
        5) type
        6) TEXT
        7) WEIGHT
        8) "2"
        9) SORTABLE
       10) NOSTEM
       11) WITHSUFFIXTRIE
```
Will WITHSUFFIXTRIE be true ? This response is from testFtInfoCommand() integration test BTW.

2. We have ongoing initiative to start promoting the Search API to production and start removing the Experimental tags. For this to happen however we needed to revisit the codecs being used in Search. It seems they are unnecessary, even worse they might introduce bugs in some scenarios. So [this PR ](https://github.com/redis/lettuce/pull/3716) aims to remove them where not applicable. However i see your PR is still using them - we need to evaluate if they are correctly used, but from first glance, i would say no.

tishun committed:
Fixing the CI test issues

tishun committed:
Fix FT.INFO parsing for RESP2/RESP3 consistency

- Fix RESP2 field attribute parsing to handle standalone boolean flags
  (SORTABLE, NOSTEM, WITHSUFFIXTRIE, etc.) which are returned as single
  elements, not key-value pairs
- Add case-insensitive handling for COORD_SYSTEM and ALGORITHM attributes
  (RESP2 returns lowercase, RESP3 returns uppercase)
- Normalize vector attribute keys to uppercase and numeric values to Long
  for consistent types across protocol versions
- Remove phonetic field from TextField as Redis FT.INFO never returns it
  even when set during index creation
- Fix test to use non-default language (FRENCH) since Redis only returns
  default_language when it differs from the default (english)
- Fix vector field DATA_TYPE attribute name (was incorrectly TYPE)

tishun committed:
Address Redis 8.8 failure

tishun committed:
Address Redis 8.8 failure 2

tishun commented:
After discussing this with @a-TODO-rov we decided to drop this solution in lue of:
- we feel the current solution is quite brittle and very likely to cause maintainence issues
- the `FT.INFO` command has an extremely complex response; which differs greatly in RESP2 and RESP3
- it might prove to be quite difficult to serve all use-cases with the same solution
- most use-cases seem to be corner cases

As such the recommneded approach would be to use [custom commands](https://redis.github.io/lettuce/advanced-usage/custom-commands/) or [command interfaces](https://redis.github.io/lettuce/redis-command-interfaces/) to achieve this for a specific use-case.

A sample solution that verifies the index needs to be recreated (for example) could be:

```java
        // On startup: introspect index schema using FT.INFO via dispatch()
        List<Object> info = redis.getStatefulConnection().sync().dispatch(CommandType.FT_INFO,
                new ArrayOutput<>(StringCodec.UTF8), new CommandArgs<>(StringCodec.UTF8).add(testIndex));

        // Extract field names from the 'attributes' section
        List<Object> attributes = (List<Object>) findInResponse(info, "attributes");
        Set<String> fieldNames = extractFieldNames(attributes);

        // Check if required fields exist - this drives migration decision
        boolean hasTitleField = fieldNames.contains("title");
        boolean hasDescriptionField = fieldNames.contains("description");

        assertThat(hasTitleField).isTrue();
        assertThat(hasDescriptionField).isFalse(); // Schema is outdated, would trigger migration
```

or

```java
// Use command interface for FT.INFO
        RedisCommandFactory factory = new RedisCommandFactory(redis.getStatefulConnection());
        FtInfoCommands cmds = factory.getCommands(FtInfoCommands.class);

        List<Object> info = cmds.ftInfo(testIndex);

        // Validate schema has all required fields
        List<Object> attributes = (List<Object>) findInResponse(info, "attributes");
        Set<String> fieldNames = extractFieldNames(attributes);

        assertThat(fieldNames).contains("title", "content"); // Schema is valid

...

/** Command interface for FT.INFO. */
    interface FtInfoCommands extends Commands {

        @Command("FT.INFO ?0")
        List<Object> ftInfo(String index);

    }
```

CC @uglide
